main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala

```diff
@@ -373,6 +374,137 @@ object Simplifier {
+  private def visitPolyType(tpe: Type): Type = {
+    val base = tpe.baseType
+    val args = tpe.typeArguments.map(visitPolyType)
+
+    base match {
+      case v@Type.Var(_, _) => Type.mkApply(v, args, tpe.loc)
+
+      case Type.Cst(tc, _) =>
+        tc match {
+          case TypeConstructor.Void => Type.Cst(tc, tpe.loc)
+
+          case TypeConstructor.AnyType => Type.Cst(tc, tpe.loc)
+
+          case TypeConstructor.Unit => Type.Cst(tc, tpe.loc)
+
+          case TypeConstructor.Null => Type.Cst(tc, tpe.loc)
+
+          case TypeConstructor.Bool => Type.Cst(tc, tpe.loc)
+
+          case TypeConstructor.Char => Type.Cst(tc, tpe.loc)
+
+          case TypeConstructor.Float32 => Type.Cst(tc, tpe.loc)
+
+          case TypeConstructor.Float64 => Type.Cst(tc, tpe.loc)
+
+          case TypeConstructor.BigDecimal => Type.Cst(tc, tpe.loc)
+
+          case TypeConstructor.Int8 => Type.Cst(tc, tpe.loc)
+
+          case TypeConstructor.Int16 => Type.Cst(tc, tpe.loc)
+
+          case TypeConstructor.Int32 => Type.Cst(tc, tpe.loc)
+
+          case TypeConstructor.Int64 => Type.Cst(tc, tpe.loc)
+
+          case TypeConstructor.BigInt => Type.Cst(tc, tpe.loc)
+
+          case TypeConstructor.Str => Type.Cst(tc, tpe.loc)
+
+          case TypeConstructor.Regex => Type.Cst(tc, tpe.loc)
+
+          case TypeConstructor.RecordRowEmpty => Type.Cst(tc, tpe.loc)
+
+          case TypeConstructor.Sender => throw InternalCompilerException("Unexpected Sender", tpe.loc)
+
+          case TypeConstructor.Receiver => throw InternalCompilerException("Unexpected Receiver", tpe.loc)
+
+          case TypeConstructor.Lazy => Type.mkApply(Type.Cst(tc, tpe.loc), args, tpe.loc)
+
+          case TypeConstructor.Enum(sym, _) => Type.mkApply(Type.Cst(tc, tpe.loc), args, tpe.loc)
+
+          case TypeConstructor.Struct(sym, _) => Type.mkApply(Type.Cst(tc, tpe.loc), args, tpe.loc)
+
+          case TypeConstructor.RestrictableEnum(sym, _) =>
+            val enumSym = new Symbol.EnumSym(sym.namespace, sym.name, sym.loc)
+            Type.mkEnum(enumSym, args, tpe.loc)
+
+          case TypeConstructor.Native(clazz) => Type.Cst(tc, tpe.loc)
+
+          case TypeConstructor.Array =>
+            // remove region
+            // OBS: non-well kinded type
+            Type.Apply(Type.Cst(TypeConstructor.Array, tpe.loc), args.head, tpe.loc)
+
+          case TypeConstructor.Vector =>
+            // OBS: non-well kinded type
+            Type.Apply(Type.Cst(TypeConstructor.Array, tpe.loc), args.head, tpe.loc)
+
+          case TypeConstructor.RegionToStar => Type.mkApply(Type.Cst(tc, tpe.loc), args, tpe.loc)
+
+          case TypeConstructor.Tuple(_) => Type.mkApply(Type.Cst(tc, tpe.loc), args, tpe.loc)
+
+          case TypeConstructor.Arrow(_) =>
+            // remove effect
+            // OBS: non-well kinded type
+            Type.mkApply(Type.Cst(TypeConstructor.Arrow(args.length - 1), tpe.loc), args.drop(1), tpe.loc)
```

Comment:
I think breaking these invariants is terrible. What can be done?

Reply:
Perhaps we need a non-Type, non-MonoType type?

Reply:
because these violations seems to extreme

Reply:
Yeah its not great. Here is a sorts of types I have:
- Monomoprh: Types without Alias, Associated types, Jvm Types (here with variables, but monomorpher would like one version of this also without variables)
- Simplifier: The above, but now also with different Kinding of array and arrows, 
- Eraser: A type that can only be a variable, a java primite, or object